src/tir/schedule/analysis/analysis.cc

```diff
@@ -1376,5 +1376,333 @@ void CheckStorageScope(const ScheduleState& self, String storage_scope) {
+  if (op->kind != rhs->kind) return false;
+  if (op->thread_binding.defined() ^ rhs->thread_binding.defined()) return false;
+  if (op->thread_binding.defined() &&
+      !VisitExpr(op->thread_binding.value(), rhs->thread_binding.value()))
+    return false;
+  return CompareAnnotationMap(op->annotations, rhs->annotations);
+}
+
+bool TensorizeComparator::VisitStmt_(const SeqStmtNode* op, const Stmt& other) {
+  const auto* rhs = other.as<SeqStmtNode>();
+  return CompareArray(op->seq, rhs->seq, &TensorizeComparator::VisitStmt);
+}
+
+bool TensorizeComparator::VisitStmt_(const BufferStoreNode* op, const Stmt& other) {
+  const auto* rhs = other.as<BufferStoreNode>();
+  return CompareBufferAccess(op, rhs) && VisitExpr(op->value, rhs->value);
+}
+
+bool TensorizeComparator::VisitStmt_(const BlockRealizeNode* op, const Stmt& other) {
+  const auto* rhs = other.as<BlockRealizeNode>();
+  // Skip Compare binding values if the block is scope block (the outermost one).
+  if (!is_scope_block) {
+    size_t offset = op->iter_values.size() - rhs->iter_values.size();
+    if (rhs->iter_values.size() > op->iter_values.size()) return false;
+    if (is_inner_block) {
+      // weak pattern matching for the inner block (the son of the scope block)
+      // where the pattern is v + iter <=> expr + iter
+      for (size_t i = 0; i < rhs->iter_values.size(); ++i) {
+        PrimExpr lhs_expr, rhs_expr;
+        Optional<Var> lhs_iter, rhs_iter;
+        auto detect = [](const PrimExpr& binding) -> std::pair<PrimExpr, Optional<Var>> {
+          arith::PVar<PrimExpr> expr;
+          arith::PVar<Var> iter;
+          if (iter.Match(binding)) {
+            return std::make_pair(0, iter.Eval());
+          } else if ((expr + iter).Match(binding)) {
+            return std::make_pair(expr.Eval(), iter.Eval());
+          } else if ((iter + expr).Match(binding)) {
+            return std::make_pair(expr.Eval(), iter.Eval());
+          } else {
+            return std::make_pair(expr.Eval(), NullOpt);
+          }
+        };
+        std::tie(lhs_expr, lhs_iter) = detect(op->iter_values[i + offset]);
+        std::tie(rhs_expr, rhs_iter) = detect(rhs->iter_values[i]);
+        CHECK((lhs_iter && rhs_iter) || (!lhs_iter && !rhs_iter)) << "Incompatible binding";
+        if (lhs_iter) VisitExpr(lhs_iter.value(), rhs_iter.value());
+        if (is_zero(rhs_expr)) {
+          CHECK(is_zero(lhs_expr)) << "Incompatible binding";
+        } else {
+          const auto* bv = rhs_expr.as<VarNode>();
+          if (!bv) {
+            VisitExpr(lhs_expr, rhs_expr);
+          } else {
+            auto it = equal_map_.find(GetRef<Var>(bv));
+            if (it == equal_map_.end()) {
+              equal_map_[GetRef<Var>(bv)] = lhs_expr;
+            } else {
+              CHECK(it->second->IsInstance<PrimExprNode>());
+              VisitExpr(lhs_expr, Downcast<PrimExpr>(it->second));
+            }
+          }
+        }
```

Comment:
we should probably restructure this a little bit

```suggestion
        } else if (const auto* bv = rhs_expr.as<VarNode>()) {
            auto it = equal_map_.find(GetRef<Var>(bv));
            if (it == equal_map_.end()) {
              equal_map_[GetRef<Var>(bv)] = lhs_expr;
            } else {
              CHECK(it->second->IsInstance<PrimExprNode>());
              VisitExpr(lhs_expr, Downcast<PrimExpr>(it->second));
            }
        } else {
            VisitExpr(lhs_expr, rhs_expr);
        }
```